src/rgw/rgw_sal_motr.cc

```diff
@@ -1923,127 +1922,130 @@ int MotrObject::MotrDeleteOp::delete_obj(const DoutPrefixProvider* dpp, optional
       if(ent.key.instance == "null")
         del_null_ref_key = true;
     }
-  } else {
-    ldpp_dout(dpp, 20) << "delete " << delete_key << " from " << tenant_bkt_name << dendl;
-    // Remove the motr object.
-    // versioning enabled and suspended case.
-    if (info.versioned()) {
-      if (source->have_instance())
+  }
+  // Remove the motr object.
+  // versioning enabled and suspended case.
+  if (info.versioned()) {
+    if (source->have_instance())
+    {
+      // delete object permanently.
+      result.version_id = ent.key.instance;
+      ldpp_dout(dpp, 20) << "delete " << delete_key << " from " << tenant_bkt_name << dendl;
+
+      if(ent.meta.size != 0)
       {
-        // delete object permanently.
-        result.version_id = ent.key.instance;
-        if (ent.meta.category == RGWObjCategory::MultiMeta)
-          rc = source->delete_part_objs(dpp);
-        else
-          rc = source->delete_mobj(dpp);
-        if (rc < 0) {
-          ldpp_dout(dpp, 0) << "Failed to delete the object from Motr. key- "<< delete_key << dendl;
+          if (ent.meta.category == RGWObjCategory::MultiMeta)
+            rc = source->delete_part_objs(dpp);
+          else
+            rc = source->delete_mobj(dpp);
+          if (rc < 0) {
+            ldpp_dout(dpp, 0) << "Failed to delete the object from Motr. key- "<< delete_key << dendl;
+            return rc;
+          }
+      }
+      // delete bucket index entry.
+      rc = source->store->do_idx_op_by_name(bucket_index_iname,
+                                            M0_IC_DEL, delete_key, bl);
+      if (rc < 0) {
+        ldpp_dout(dpp, 0) << "Failed to del object's entry " << delete_key << "from bucket index. " << dendl;
+        return rc;
+      }
+      // if deleted object version is the latest version,
+      // then update is-latest flag to true for previous version.
+      if(ent.is_current())
+      {
+        ldpp_dout(dpp, 20)<<__func__<< "Updating previous version entries " << dendl;
+        bool set_is_latest=true;
+        rc = source->update_version_entries(dpp, set_is_latest);
+        if (rc < 0)
           return rc;
-        }
-        // delete bucket index entry.
-        rc = source->store->do_idx_op_by_name(bucket_index_iname,
-                                              M0_IC_DEL, delete_key, bl);
-        if (rc < 0) {
-          ldpp_dout(dpp, 0) << "Failed to del object's entry " << delete_key << "from bucket index. " << dendl;
+      }
+      // delete null reference key
+      if(ent.key.instance == "null")
+        del_null_ref_key = true;
+    }
+    else{
+      // generate version-id for delete marker.
+      result.delete_marker = true;
+      source->gen_rand_obj_instance_name();
+
+      // if latest version is null version, then delete the null version-object and
+      // add reference of delete-marker in null reference key.
```

Comment:
In case of suspended bucket scenarios, at a time there will be only one null entry, it can be null version or null delete marker.

- When we delete an object in suspended bucket, we add a null delete marker and add it's reference to null entry(obj/null). Also, we delete the previous null version object
- Hence, we require to delete that entry when we delete the delete marker.

This code requires some optimization which we will cover as a part of another JIRA - https://jts.seagate.com/browse/CORTX-31978

Reply:
There is nothing called "null delete marker". When we delete an object in suspended bucket, we should straight away delete the existing null object if present when version id is not specified in req. We should confirm this behavior on AWS and Ceph.